R/process_source.R

```diff
@@ -1,131 +1,71 @@
-        # starts here.
-        state <- "OBJECT_DEFINITION"
-        current_chunk$add(line)
-        current_chunk_name <- extract_object_name(line)
-
-        # Special case: check if it's a one-liner (one-liners will be parseable;
-        # an opening line for multi-line functions/expressions will not.). If
-        # so, transition to SCANNING state.
-        if (is_parseable(line)) {
-          state <- "SCANNING"
-          chunks[[current_chunk_name]] <- current_chunk$get()
-          current_chunk$reset()
-          current_chunk_name <- NULL
-        }
-
-      } else if (grepl("^#", line)) {
-        state <- "LEADING_COMMENT"
-        current_chunk$add(line)
-      }
-
-    } else if (state == "OBJECT_DEFINITION") {
-      current_chunk$add(line)
-      if (grepl("^\\S", line)) {
-        # If the line starts with anything _other_ than whitespace, then we know
-        # we've reached the end of the object definition.
-        state <- "SCANNING"
-        chunks[[current_chunk_name]] <- current_chunk$get()
-        current_chunk$reset()
-        current_chunk_name <- NULL
-      }
-
-    } else if (state == "LEADING_COMMENT") {
-      if (grepl("^\\S+ *<-", line)) {
-        # Lines that start with "foo <-" tell us that the object definition
-        # starts here.
-        state <- "OBJECT_DEFINITION"
-        current_chunk$add(line)
-        current_chunk_name <- extract_object_name(line)
-
-        # Special case: check if it's a one-liner (one-liners will be parseable;
-        # an opening line for multi-line functions/expressions will not.). If
-        # so, transition to SCANNING state.
-        if (is_parseable(line)) {
-          state <- "SCANNING"
-          chunks[[current_chunk_name]] <- current_chunk$get()
-          current_chunk$reset()
-          current_chunk_name <- NULL
-        }
-
-      } else if (grepl("^#", line)) {
-        # Stay in current state.
-        current_chunk$add(line)
-
-      } else {
-        # We've hit the end of a comment block but no definition here. Just
-        # discard the block and go back to scanning mode.
-        state <- "SCANNING"
-        current_chunk$reset()
-      }
-    }
+# Given a file, process the text and return a named list of objects.
+process_source_file_one <- function(file) {
+  parse_data <- utils::getParseData(parse(file = file, keep.source = TRUE))
+  text <- readLines(file)
+
+  assignment_ops <- parse_data[
+    parse_data$token %in% c("LEFT_ASSIGN", "EQ_ASSIGN", "RIGHT_ASSIGN"),
+  ]
+  # Assignments are the parent expressions of assignment operators
+  assignments <- parse_data[parse_data$id %in% assignment_ops$parent, ]
+  # Top-level expressions have a parent attribute of 0
+  object_definitions <- assignments[assignments$parent == 0, ]
+
+  staticexports <- vector("list", nrow(object_definitions))
+
+  for (i in seq_len(nrow(object_definitions))) {
+    object_definition <- object_definitions[i, ]
```

Comment:
This is minor, but I don't like iterating over rows of the data frame and slicing out a single row each time, because, it's slow. (Creating data frames is slow in general in R.)

Also, using a value from the one-row data frame and relying on the fact that's it's a one-element vector (as in `object_definitions$id` instead of `object_definitions$id[1]`), feels weird to me.

Can you pull out the necessary values into a list instead? With a list, the shape of the data is clearer, and it will be faster.

Reply:
I did a refactor in [47370ae](https://github.com/wch/staticimports/pull/21/commits/47370ae48ac2d72d079fbff2c2111a741a4a5c78) that I'm hoping should address this. I didn't know about the performance penalties for slicing rows from data frames, so that'll be good to keep in mind for the future.